sdk/trace/batch_span_processor.go

```diff
@@ -280,6 +280,7 @@ func (bsp *batchSpanProcessor) exportSpans(ctx context.Context) error {
 		//
 		// It is up to the exporter to implement any type of retry logic if a batch is failing
 		// to be exported, since it is specific to the protocol and backend being sent to.
+		clear(bsp.batch) // Let GC collect objects
```

Comment:
I think that not clearing via `clear(bsp.batch)` and doing `bsp.batch = bsp.batch[:0]` below is intentional to reuse memory and reduce number of heap allocation.

Reply:
There are two kinds of memory reuse.

If we did `bsp.batch = nil` that would indeed throw away the slice's backing array. Instead, we reuse that backing array by re-slicing the slice to have zero length but keeping the capacity and array  (`bsp.batch[:0]`).

`clear()` does not change the above.

What `clear()` does is it erases the backing array elements (from zero to `len()`) with zero values for the array type. IIRC (I'm not looking at the code right now) the type is an interface. So, instead of pointers to objects, we have `nil`s there. And, hence, GC can collect those objects if they are not referenced anywhere anymore.

If the slice was of a scalar type (e.g. `int`), then `clear()` would not do anything useful for GC. But for interfaces it does.

I hope this helps to understand why I think this change would be beneficial. Or did I misunderstand your concern?

Reply:
Makes sense. Can you `benchstat` put results to PR description to compare the performance difference caused by the change?

Reply:
Added a benchmark to the description.

Reply:
@pellared I've looked at other `[:0]` usage and made a few more improvements. PTAL.